### PR TITLE
Federation

### PR DESCRIPTION
**Pull request by sedkis**
Adds Federation example for v4.0.0.  

See new README for details.

### CODE REVIEW

**Comment by sedkis** [comment]
@rewsmith @davegarvey  help me understand how the tests work.  

I added a test to the main deployment postman collection.  It gets executed, but fails, because the `federation` deployment isn't run.

I was under the impression that all the deployments will get spun up - but this seems NOT to be the case.

I can see that some of the deployments have their own Postman collections.  Are these run as part of the CICD tests?

**Comment by davegarvey** [comment]
@sedkis answers below...

> I added a test to the main deployment postman collection. It gets executed, but fails, because the federation deployment isn't run.

Only the main `tyk` deployment is deployed, so tests based on other deployments will fail.

Tests related to a particular deployment should be added to a postman collection contained within that deployment, not the main collection.

> I was under the impression that all the deployments will get spun up - but this seems NOT to be the case.

Only the `tyk` deployment is deployed during testing. The test config [runs the `./up.sh` command without any additional deployment arguments](https://github.com/TykTechnologies/tyk-demo/blob/faab438c90d5cb37a5211b0d38823814be988905/.github/workflows/tyk-demo-tests.yml#L21):

```
        run: sudo ./up.sh
```

I suppose this is as a result of this being an MVP approach. Perhaps it's now time to improve things.

> I can see that some of the deployments have their own Postman collections. How are these tested?

Yes, I separated the postman collections across the deployments, as they only work in the context of their own deployment. Their tests would fail if they were added as part of the main `tyk` postman collection.

The non-`tyk` deployments aren't currently tested automatically. The [test script only references the `tyk` deployment postman collection](https://github.com/TykTechnologies/tyk-demo/blob/faab438c90d5cb37a5211b0d38823814be988905/scripts/test.sh#L21):

```
docker run -t --rm \
    --network tyk-demo_tyk \
    -v $(pwd)/deployments/tyk/tyk_demo.postman_collection.json:/etc/postman/tyk_demo.postman_collection.json \
    -v $(pwd)/test.postman_environment.json:/etc/postman/test.postman_environment.json \
    postman/newman:alpine \
    run "/etc/postman/tyk_demo.postman_collection.json" \
    --environment /etc/postman/test.postman_environment.json \
    --insecure
```

> if the tests do NOT get run for the deployment specific collections automatically, are the expectations to run them manually before creating PRs?

Yes. But we could look at automating the process, so that all deployments are deployed and tested. 

However, it not a totally straight forward process. I expect that we'd have to orchestrate the deployments to avoid issues or incompatibilities with deploying and testing everything in one go. This is because some deployments remove or alter services for their own needs, which would likely cause problems with other deployments - hence this "deployment" approach being required.

Solution could be to create some logic to loop through the deployments, but taking into account that some deployments rely on others e.g. `cicd` relies on `tyk2`. It would increase the time required to run the tests, but would give full coverage. This would be preferable to the current situation though.

File: deployments/federation/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+1. Once running, visit the Supergraph playground:
+
+http://tyk-gateway.localhost:8080/social-media-apis-federated/playground
+
+2. The example includes Users and Posts subgraph.  
+
+3. Add the "comments" subgraph by adding a new Tyk API Subgraph:
+```
+http://comments-subgraph:4003/query
```
Comment (JRWu) [inline deployments/federation/README.md]: Should change the URL to:
http://comments:4003/query
Verified that it works

**Review by JRWu — APPROVED** [review]
Work looks mint to me, pulled on my local pc & tested the federation example- worked great.

**Pull request merged**